calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+            && Self::guard_true(&assign.guard)
+            && Self::src_const(&assign.src)
+        {
+            if let ir::PortParent::Cell(cell_wref) = &dst_ref.parent {
+                return Some(Rc::clone(&cell_wref.upgrade()));
+            }
+        }
+        None
+    }
+
+    //returns true if port is 1'd1
+    fn src_const(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard is True
+    fn guard_true(guard: &ir::Guard) -> bool {
+        matches!(guard, ir::Guard::True)
+    }
+
+    //given a vec of assignments, return true if name is ever written to in
+    //assignments, false otherwise
+    fn search_assignments(&self, assignments: &[ir::Assignment]) -> bool {
+        for write in assignments.iter().filter_map(Self::guaranteed_go) {
+            if *write.borrow().name() == self.name {
+                return true;
+            }
+        }
+        false
+    }
+
+    //Returns true if any of the control statements in dominators write to a cell
+    //with self's name.
+    fn is_written(
+        &self,
+        dominators: &HashSet<u64>,
+        comp: &mut ir::Component,
+        shareable: &ShareSet,
+    ) -> bool {
+        for dominator in dominators {
+            if let Some(c) =
+                DominatorMap::get_control(*dominator, &comp.control.borrow())
+            {
+                match c {
+                    ir::Control::Empty(_)
+                    | ir::Control::Seq(_)
+                    | ir::Control::Par(_) => {
+                        unreachable!(
+                            "no empty/seqs/pars should be in domination map"
+                        )
+                    }
+                    ir::Control::Enable(ir::Enable { group, .. }) => {
+                        if self.search_assignments(&group.borrow().assignments)
+                        {
+                            return true;
+                        }
+                    }
+                    //You can't have a write to a stateful component in a
+                    //combinational group.
+                    ir::Control::While(_) | ir::Control::If(_) => (),
+                    ir::Control::Invoke(ir::Invoke {
+                        comp,
+                        inputs,
+                        outputs,
+                        ..
+                    }) => {
+                        for (_, port) in outputs.iter() {
+                            if self.parent_matches_name(port) {
+                                return true;
+                            }
+                        }
+                        if shareable.is_shareable_component(comp)
+                            && !inputs.is_empty()
+                        {
+                            return *comp.borrow().name() == self.name;
+                        }
+                    }
+                }
+            } else {
+                unreachable!("should always be able to get control from id")
```

Comment:
If this unreachable then change the `if let` into a `.get(..).unwrap_or_else(|| unreachable!(..))` call above